File: dataset/domain/src/main/resources/META-INF/resources/javascript/domain/dimension.js
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
 */

/**
 * Dimension
 * 
 * @author levk
 * @since CRYSTAL
 */
function Dimension (_dataset, _dimension) {
  this._dataset = _dataset;
  this._dimension = _dimension;

  /**
   * Name of the dimension (for example "row")
   */
  this.name = function () {
    return this._dataset.dimensions[_dimension].name;
  };

  /**
   * Keys of the dimension
   */
  this.keys = function () {
    return this._dataset.dimensions[_dimension].keys;
  };

  /**
   * @param _filter
   *          to return true/false for each key in the dimension
   * @param _function
   *          to be called for each filtered value, result of the function, if
   *          not null will be collected and returned in a list
   */
  this.apply = function (_filter, _function) {
    var _results = [];
    for (_key in this.keys ())
      if (_filter (this.keys ()[_key])) for (_value in this._dataset.values)
        if (this._dataset.values[_value].coordinates[this.name ()] === this.keys ()[_key]) {
          _result = _function (this._dataset.values[_value]);
          if (_result != null) _results.push (_result);
        }
    return _results;
  }
}

File: dataset/domain/src/main/java/edu/dfci/cccb/mev/dataset/domain/Dimension.java
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
 */

package edu.dfci.cccb.mev.dataset.domain;

import static javax.xml.bind.annotation.XmlAccessType.NONE;

import java.util.Iterator;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;

/**
 * Dimension
 * 
 * @author levk
 * @since BAYLIE
 */
@XmlRootElement
@XmlAccessorType (NONE)
public interface Dimension <K> extends Iterable<K>, Named {

  /**
   * Identifier for injection
   */
  public static final String DIMENSION = "dimension";

  /* (non-Javadoc)
   * @see java.lang.Iterable#iterator() */
  @Override
  @XmlElement (name = "keys")
  Iterator<K> iterator ();

  /**
   * @return number of keys
   */
  int size ();

  /**
   * @param index
   * @return key at index specified
   */
  @Path ("/{index}")
  @GET
  K get (@PathParam ("index") int index);
}
